### DISCUSSION TITLE
Normalising IPDF Triangles

### DISCUSSION THREAD

**Discussion by miamaric12** [mtex-toolbox/mtex#525] [Ask Anything]
Hi, I was wondering if someone could help me. I have segmented data that I would like to extract some information from, but i would like to normalise it over the total IPDF, is this possible in MTEX? Like in this case, I would like to normalise each of the individual  IPDF triangles over the Total IPDF triangle :) 

Thanks for the help. 
Mia
<img width="1048" alt="Screenshot 2020-10-15 at 10 20 22" src="https://user-images.githubusercontent.com/49954369/96104219-7858bd00-0ed0-11eb-9f5c-87f84076c928.png">

**Comment by kilir** [comment]
Hi Mia,
could you explain in more detail what you mean by normalizing to the entire IPDF? Take each distribution (e.g. the one you labeled "Basal Slip") and divide it by the one you labeled "Full IPDF"?
Cheers,
Rüdiger

**Reply by miamaric12** [reply]
Yes this is exactly what i want to do, I run into a problem because each of these distributions are not equal in size to the Full IPDF

Thank you for the help :) 
Mia

**Comment by kilir** [comment]
If you calculate them as spherical functions (class S2Fun) you can simply divide one by the other. Here's an example of taking some misorientation axes and normalizing their distribution by the distribution of misorientation axes induced by the crystal symmetry,.
```matlab
cs=crystalSymmetry('222')
% symmetry induced axis distribution
dist1 = calcAxisDistribution(cs,cs);

% make up some data
mdf = calcDensity(orientation.rand(5,cs,cs),'halfwidth',15*degree)
mori = discreteSample(mdf,400)

% option one
dist2 = calcDensity(mori.axis)
dist2.antipodal=1;
% normalize
dist1_norm_by_dist2 = dist1./dist2;
dist1_norm_by_dist2 = dist1_norm_by_dist2*(1/dist1_norm_by_dist2.mean);

%
plot(dist1)
mtexTitle('symmetry induced')
nextAxis
plot(dist2)
mtexTitle(' "measured" ')
nextAxis
plot(dist1_norm_by_dist2)
mtexTitle('normalized')
mtexColorbar
% note that the first two plot are mud while the third one is something
% like multiples of dist1
```
<img width="991" alt="naxd" src="https://user-images.githubusercontent.com/11697557/96121056-ccbf6500-0eef-11eb-9caa-3c960652eb83.png">

So, as long as you can compute a spherical function from your data, this should not be a problem.
Cheers,
Rüdiger

**Comment by miamaric12** [comment]
**Hi, 
Thank you so much for the help. I have been working in this, but am still abit stuck here is my code below:** 

white_odf = calcDensity(ebsd_good('Indexed').orientations(whiteInd),'resolution',10*degree)
odf_full = calcODF(ebsd_full(phase_of_interest).orientations,'halfwidth',half_width*degree)
% option one
dist1 = calcAxisDistribution(white_odf);
dist2 = calcDensity(odf_full.CS.axes)
dist2.antipodal=1;
% normalize
dist1_norm_by_dist2 = dist1./dist2;
dist1_norm_by_dist2 = dist1_norm_by_dist2*(1/dist1_norm_by_dist2.mean);

plot(dist1_norm_by_dist2)
mtexTitle('normalized')
mtexColorbar

**When i try to plot it I get this:** 
<img width="1218" alt="Screenshot 2020-10-26 at 16 17 28" src="https://user-images.githubusercontent.com/49954369/97198483-c01bf600-17a6-11eb-9240-3b1a95ca0ddb.png">

**I would like it to be plotted as an IPF triangle, but when i try to do this i get this error:** 
Error using plotIPDF
Using plotIPDF with EBSD data is obsolete since MTEX 4.0. Use plotIPDF(ebsd.orientations) instead.
but dist1_norm_by_dist2 doesnt have any orientations so i am not sure what to do. 

**Thank you so much for the help :) 
Mia**

**Comment by kilir** [comment]
Hi,
my example was for misorientation axes, but you can do this for whatever density distribution as well. I think there are quite some things which I do not understand in the lines you wrote. You compute the axis distribution of your odf and normalize by the density function obtained from the axes of your symmetry elements - which are of type vector3d (so, no symmetry here). Is this really what you want?

If your result would have been of type S2FunHarmonicSym, Mtex should have plotted only the fundamental region.

Maybe you should describe in more detail what you'd like to do.

Cheers,
Rüdiger

**Reply by kilir** [reply]
Here's another example of a density function of crystal directions parallel to 0,0,1 specimen direction ("ipf with respect to z") normalized by some other density function.
```matlab
% an odf
mtexdata dubna
odf = pf.calcODF;
%
% calc a densityfunction e.g. inverse polefigure wrt 0,0,1
ipfZ = calcPDF(odf,[],zvector)

% some othe function
distn= calcDensity(Miller(0,0,0,1,odf.CS),'halfwidth',30*degree)

% normalize
distN = ipfZ./distn;
distN = distN*(1/distN.mean);

%
plot(ipfZ)
mtexTitle('IPFZ')
nextAxis
plot(distn)
mtexTitle('a function used for normalization')
nextAxis
plot(distN)
mtexTitle('result')
mtexColorbar
```

**Comment by miamaric12** [comment]
**Thank you so much for the help :) 
Ok i will try and explain it from the beginning please let me know if i am still being unclear**

**The first thing that i have done is Calculate the misorientation axis & plot in the crystal frame:**
```matlab
% Calculate the misorientation in the specimen frame
axis_specimen = axis(grains_clean(ebsd_full(phase_of_interest).grainId).meanOrientation,...
  ebsd_full(phase_of_interest).orientations);
% Calculate the misorientation in the crystal frame (i.e. rotate each according to the grain mean orientation
axis_crystal = axis(inv(grains_clean(ebsd_full(phase_of_interest).grainId).meanOrientation) .* ...
  ebsd_full(phase_of_interest).orientations);
angle_crystal = angle(inv(grains_clean(ebsd_full(phase_of_interest).grainId).meanOrientation) .* ... 
  ebsd_full(phase_of_interest).orientations);
```
This gives a plot that looks something like this: 
![image](https://user-images.githubusercontent.com/49954369/97207839-5bff2f00-17b2-11eb-9fcc-959579772f65.png)

I want to separate all of the colours in the pictures so i did this by colour segmentation and plotted the colours in IPF triangles to visualise as shown in this image below: 
<img width="973" alt="Screenshot 2020-10-26 at 17 45 05" src="https://user-images.githubusercontent.com/49954369/97208390-feb7ad80-17b2-11eb-905d-d055232b9784.png">
 where the different colours were oriented: 
```matlab
for ii = 1:1:length(col1alys)
    if col1alys(ii,1) > colorLimit && col1alys(ii,2) > colorLimit && col1alys(ii,3) > colorLimit % 6
        col1alys(ii,4) = 7;
    end
end
% Find blue pixels 
whiteInd = find(col1alys(:,4) == 7);
% Plot 
figure(14);
plotIPDF(ebsd_good('Indexed').orientations(whiteInd), RGB_reduced(whiteInd,1:3), ...
    [yvector], 'points','all', 'MarkerSize',4,...
    'MarkerEdgeColor','k', 'antipodal'); 
title(['Kind of White-ish , Amount(%) = ', num2str(length(whiteInd)/length(col1alys)*100)], 'FontSize', 18)
disp(['White = ', num2str(length(whiteInd)/length(col1alys)*100)])
```

After doing this, to better understand which orientations the colours were at the highest density (this gives us an idea about which slip mechanism is occurring) i plotted the IPDF triangles shown in the first post
```matlab
white_odf = calcDensity(ebsd_good('Indexed').orientations(whiteInd),'resolution',10*degree)
```
What I wanted to do was normalize the different colour orientations over the full orientation so that i will be able to better see trends. I want to do this by plotting the IPDF triangles of the normalised data if that makes sense :) 

Thank you again for the help. Sorry if this is confusing..

**Reply by kilir** [reply]
I guess, I see.
Question 1: What do you mean by normalizing "over the full orientation"?

Question 2: What are the colors in your map? Assuming that it is some colorcoding for crystal directions, are these your misorientation axes or are these your crystals direction being parallel to one common specimen direction ("IPF colorcoding")? 

Comment: I think in both cases for question 2, there are easier ways to find those pixels instead of segmenting the colors.
Another comment: I think to deduce slip systems from the orientation to the mean orientation requires at least very great care. Under some (most ?) circumstances, it is probably not possible - at least not physically related to your crystal.

Cheers,
Rüdiger
